### PR TITLE
esdoc Dependencies

### PR DESCRIPTION
**Pull request by vankop**
https://github.com/webpack/webpack/issues/3277

**What kind of change does this PR introduce?**

esdoc for Dependency, NullDependency, HarmonyExportExpressionDependency, HarmonyCompatibilityDependency, HarmonyExportHeaderDependency, HarmonyInitDependency, HarmonyExportSpecifierDependency

**Did you add tests for your changes?**

No

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

Nothing

### CODE REVIEW

File: lib/dependencies/HarmonyExportExpressionDependency.js [hunk 1]
```diff
@@ -26,7 +56,15 @@ class HarmonyExportExpressionDependency extends NullDependency {
 	}
 }
 
+/**
+ * @implements {DependencyTemplate}
```
Comment (vankop) [inline lib/dependencies/HarmonyExportExpressionDependency.js]: Unfortunately, `@implements` does not work for Typescript yet, and workaround with `{{prototype: DependencyTemplate}}` does not as well. Workaround works only for definition (object still is not callable, fails on `new Template()`). So I implemented each method to support Typescript in templates.

Comment (sokra) [reply lib/dependencies/HarmonyExportExpressionDependency.js]: You could create a `DependencyTemplate` class and `extends DependencyTemplate` for all dep templates

File: lib/Dependency.js [hunk 2]
```diff
@@ -71,6 +72,10 @@ class Dependency {
 		return null;
 	}
 
+	/**
+	 * @param {Hash} hash - hash
```
Comment (sokra) [inline lib/Dependency.js]:
```suggestion
	 * @param {Hash} hash hash
```

Don't put dashes here.

File: lib/dependencies/HarmonyCompatibilityDependency.js [hunk 3]
```diff
@@ -3,20 +3,41 @@
...
+	/**
+	 * @override
+	 */
 	get type() {
 		return "harmony export header";
 	}
 }
 
+/**
+ * @implements {DependencyTemplate}
```
Comment (sokra) [inline lib/dependencies/HarmonyCompatibilityDependency.js]: If this `@implements` doesn't work, we should forbid the usage in `.eslintrc.js`.

File: lib/dependencies/HarmonyInitDependency.js [hunk 4]
```diff
@@ -4,39 +4,74 @@
...
+		 */
 		const list = [];
 		for (const dependency of module.dependencies) {
 			const template = dependencyTemplates.get(dependency.constructor);
+
 			if (
 				template &&
-				typeof template.harmonyInit === "function" &&
-				typeof template.getHarmonyInitOrder === "function"
+				typeof /** @type {HarmonyDependencyTemplate} */ (template).harmonyInit ===
```
Comment (sokra) [inline lib/dependencies/HarmonyInitDependency.js]: Only cast it once. Maybe in or after line 57.

File: lib/Dependency.js (hunk 2 above)
Comment (vankop) [reply lib/Dependency.js]: https://github.com/gajus/eslint-plugin-jsdoc#eslint-plugin-jsdoc-rules-require-hyphen-before-param-description

File: lib/dependencies/HarmonyInitDependency.js (hunk 4 above)
Comment (vankop) [reply lib/dependencies/HarmonyInitDependency.js]: done

File: lib/dependencies/HarmonyCompatibilityDependency.js (hunk 3 above)
Comment (vankop) [reply lib/dependencies/HarmonyCompatibilityDependency.js]: done in https://github.com/webpack/webpack/commit/a1aa5cbdce7110ab18aba098492853150a19aa0b

File: lib/Dependency.js (hunk 2 above)
Comment (vankop) [reply lib/Dependency.js]: done

File: lib/dependencies/HarmonyExportExpressionDependency.js (hunk 1 above)
Comment (vankop) [reply lib/dependencies/HarmonyExportExpressionDependency.js]: done

**Comment by vankop** [comment]
Looks like this already done in `next` branch..

**Pull request closed without merging**